// @ymjacky/mqtt5/lib/cache/lru_cache.js
// least recentry used cache
var _computedKey;
_computedKey = Symbol.iterator;
export class LruCache {
  _capacity;
  _maxAge;
  _values;
  _ages;
  constructor(capacity, maxAge = Number.MAX_VALUE){
    this._capacity = capacity;
    this._maxAge = maxAge;
    this._values = new Map();
    this._ages = new Map();
  }
  *[_computedKey]() {
    for (const val of this._values.entries()){
      yield val;
    }
  }
  capacity() {
    return this._capacity;
  }
  lruKey() {
    const sortedByValueAsc = new Map([
      ...this._ages
    ].sort((a, b)=>a[1] - b[1]));
    const firstEntry = sortedByValueAsc.keys().next();
    const key = firstEntry.value;
    return key;
  }
  set(key, value) {
    if (!this._values.has(key)) {
      if (this._ages.size == this._capacity) {
        const key = this.lruKey();
        if (key) this.delete(key);
      }
    }
    this._values.set(key, value);
    this._ages.set(key, 1);
  }
  get(key) {
    const value = this._values.get(key);
    if (value) {
      let age = this._ages.get(key);
      if (age) {
        age++;
        if (age > this._maxAge) {
          this._values.delete(key);
          this._ages.delete(key);
        } else {
          this._ages.set(key, age);
        }
      } else {
        this._ages.set(key, 1);
      }
    }
    return value;
  }
  clear() {
    this._ages.clear();
    this._values.clear();
  }
  delete(key) {
    this._ages.delete(key);
    return this._values.delete(key);
  }
  size() {
    return this._values.size;
  }
  has(key) {
    return this._values.has(key);
  }
  keys() {
    return this._values.keys();
  }
  values() {
    return this._values.values();
  }
  entries() {
    return this._values.entries();
  }
}
