// @ymjacky/mqtt5/lib/mqtt_packets/mod.ts
/**
 * This module contains definitions and functions for MQTT packets and functions for MQTT packets
 * @module
 */

import * as MqttUtils from '../mqtt_utils/mod.js';
import type { ConnectPacket } from './connect.js';
import { parse as connectParse, toBytes as connectToBytes } from './connect.js';
import type { ConnackPacket } from './connack.js';
import { parse as connackParse, toBytes as connackToBytes } from './connack.js';
import type { PublishPacket } from './publish.js';
import { parse as publishParse, toBytes as publishToBytes } from './publish.js';
import type { PubackPacket } from './puback.js';
import { parse as pubackParse, toBytes as pubackToBytes } from './puback.js';
import type { PubrecPacket } from './pubrec.js';
import { parse as pubrecParse, toBytes as pubrecToBytes } from './pubrec.js';
import type { PubrelPacket } from './pubrel.js';
import { parse as pubrelParse, toBytes as pubrelToBytes } from './pubrel.js';
import type { PubcompPacket } from './pubcomp.js';
import { parse as pubcompParse, toBytes as pubcompToBytes } from './pubcomp.js';

import type { SubscribePacket, Subscription } from './subscribe.js';
import { parse as subscribeParse, toBytes as subscribeToBytes } from './subscribe.js';
import type { SubackPacket } from './suback.js';
import { parse as subackParse, toBytes as subackToBytes } from './suback.js';
import type { UnsubscribePacket } from './unsubscribe.js';
import { parse as unsubscribeParse, toBytes as unsubscribeToBytes } from './unsubscribe.js';
import type { UnsubackPacket } from './unsuback.js';
import { parse as unsubackParse, toBytes as unsubackToBytes } from './unsuback.js';
import type { PingreqPacket } from './pingreq.js';
import { parse as pingreqParse, toBytes as pingreqToBytes } from './pingreq.js';
import type { PingrespPacket } from './pingresp.js';
import { parse as pingrespParse, toBytes as pingrespToBytes } from './pingresp.js';
import type { DisconnectPacket } from './disconnect.js';
import { parse as disconnectParse, toBytes as disconnectToBytes } from './disconnect.js';
import type { AuthPacket } from './auth.js';
import { parse as authParse, toBytes as authToBytes } from './auth.js';

import * as Mqtt from '../mqtt/mod.js';
import * as MqttPacketsError from './error.js';
export * as MqttPacketsError from './error.js';

export type { Subscription };

export type AnyPacket =
  | ConnectPacket
  | ConnackPacket
  | PublishPacket
  | PubackPacket
  | PubrecPacket
  | PubrelPacket
  | PubcompPacket
  | SubscribePacket
  | SubackPacket
  | UnsubscribePacket
  | UnsubackPacket
  | PingreqPacket
  | PingrespPacket
  | DisconnectPacket
  | AuthPacket;

export type {
  AuthPacket,
  ConnackPacket,
  ConnectPacket,
  DisconnectPacket,
  PingreqPacket,
  PingrespPacket,
  PubackPacket,
  PubcompPacket,
  PublishPacket,
  PubrecPacket,
  PubrelPacket,
  SubackPacket,
  SubscribePacket,
  UnsubackPacket,
  UnsubscribePacket,
};

export function packetToBytes(
  packet: AnyPacket,
  protocolVersion: Mqtt.ProtocolVersion,
): Uint8Array {
  try {
    let bytes: Uint8Array;
    switch (packet.type) {
      case 'connect':
        bytes = connectToBytes(packet);
        break;
      case 'connack':
        bytes = connackToBytes(packet, protocolVersion);
        break;
      case 'publish':
        bytes = publishToBytes(packet, protocolVersion);
        break;
      case 'puback':
        bytes = pubackToBytes(packet, protocolVersion);
        break;
      case 'pubrec':
        bytes = pubrecToBytes(packet, protocolVersion);
        break;
      case 'pubrel':
        bytes = pubrelToBytes(packet, protocolVersion);
        break;
      case 'pubcomp':
        bytes = pubcompToBytes(packet, protocolVersion);
        break;
      case 'subscribe':
        bytes = subscribeToBytes(packet, protocolVersion);
        break;
      case 'suback':
        bytes = subackToBytes(packet, protocolVersion);
        break;
      case 'unsubscribe':
        bytes = unsubscribeToBytes(packet, protocolVersion);
        break;
      case 'unsuback':
        bytes = unsubackToBytes(packet, protocolVersion);
        break;
      case 'pingreq':
        bytes = pingreqToBytes(packet);
        break;
      case 'pingresp':
        bytes = pingrespToBytes(packet);
        break;
      case 'disconnect':
        bytes = disconnectToBytes(packet, protocolVersion);
        break;
      case 'auth':
        bytes = authToBytes(packet, protocolVersion);
        break;
      default:
        throw new MqttPacketsError.UnsuportedPacketError('unknown');
    }

    return bytes;
  } catch (err) {
    throw new MqttPacketsError.MqttPacketSerializeError('mqtt packet serialize error.', { cause: err });
  }
}

export function decode(
  buffer: Uint8Array,
  protocolVersion: Mqtt.ProtocolVersion = Mqtt.ProtocolVersion.MQTT_V5,
): AnyPacket {
  let pos = 0;

  try {
    // parse packetType
    const packetType = buffer[pos] >> 4;
    const packetFlags = buffer[pos++] & 0b00001111;

    // parse remainingLength
    const remainingLengthBytes = [];
    let readLength = 0;
    do {
      const readValue = buffer[pos++];
      readLength++;

      remainingLengthBytes.push(readValue);

      if ((readValue >> 7) == 0) {
        break;
      } else if (readLength == 4) {
        throw new Error('malformed packet');
      }
    } while (readLength < 4);

    const remainingLength = MqttUtils.variableByteIntegerToNum(
      new Uint8Array([...remainingLengthBytes]),
      0,
    );

    let variableHeaderAndPayload = new Uint8Array(0);
    if (remainingLength.number > 0) {
      variableHeaderAndPayload = new Uint8Array(buffer.slice(pos));
    }

    const packet = parsePacket(
      packetType,
      packetFlags,
      remainingLength.number,
      variableHeaderAndPayload,
      protocolVersion,
    );
    return packet;
  } catch (err) {
    throw new MqttPacketsError.MqttPacketParseError('mqtt packet parse error.', { cause: err });
  }
}

function parsePacket(
  packetTypeId: number,
  packetFlags: number,
  remainingLength: number,
  variableHeaderAndPayload: Uint8Array,
  protocolVersion: Mqtt.ProtocolVersion,
): AnyPacket {
  let packet;

  switch (packetTypeId) {
    case Mqtt.PacketType.CONNECT:
      packet = connectParse(variableHeaderAndPayload, remainingLength);
      break;
    case Mqtt.PacketType.CONNACK:
      packet = connackParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.PUBLISH:
      packet = publishParse(
        packetFlags,
        variableHeaderAndPayload,
        remainingLength,
        protocolVersion,
      );
      break;
    case Mqtt.PacketType.PUBACK:
      packet = pubackParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.PUBREC:
      packet = pubrecParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.PUBREL:
      packet = pubrelParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.PUBCOMP:
      packet = pubcompParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.SUBSCRIBE:
      packet = subscribeParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.SUBACK:
      packet = subackParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.UNSUBSCRIBE:
      packet = unsubscribeParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.UNSUBACK:
      packet = unsubackParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.PINGREQ:
      packet = pingreqParse(remainingLength);
      break;
    case Mqtt.PacketType.PINGRESP:
      packet = pingrespParse(remainingLength);
      break;
    case Mqtt.PacketType.DISCONNECT:
      packet = disconnectParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    case Mqtt.PacketType.AUTH:
      packet = authParse(variableHeaderAndPayload, remainingLength, protocolVersion);
      break;
    default:
      throw new MqttPacketsError.UnsuportedPacketError('unknown');
  }

  return packet as AnyPacket;
}
