// @ymjacky/mqtt5/lib/mqtt_packets/unsuback.js
import * as Mqtt from "../mqtt/mod.js";
import { numToTwoByteInteger, numToVariableByteInteger, twoByteIntegerToNum, variableByteIntegerToNum } from "../mqtt_utils/mod.js";
import * as MqttProperties from "../mqtt_properties/mod.js";
export function toBytes(packet, protocolVersion) {
  const variableHeader = [
    ...numToTwoByteInteger(packet.packetId)
  ]; //packetId
  if (protocolVersion > Mqtt.ProtocolVersion.MQTT_V3_1_1) {
    // MQTT V5.0
    variableHeader.push(...MqttProperties.propertiesToBytes(packet.properties));
  }
  if (protocolVersion > Mqtt.ProtocolVersion.MQTT_V3_1_1) {
    const payload = packet.reasonCodes;
    const fixedHeader = [
      (Mqtt.PacketType.UNSUBACK << 4) + Mqtt.fixexHeaderFlag.UNSUBACK,
      ...numToVariableByteInteger(variableHeader.length + payload.length)
    ];
    return Uint8Array.from([
      ...fixedHeader,
      ...variableHeader,
      ...payload
    ]);
  } else {
    const fixedHeader = [
      (Mqtt.PacketType.UNSUBACK << 4) + Mqtt.fixexHeaderFlag.UNSUBACK,
      ...numToVariableByteInteger(variableHeader.length)
    ];
    return Uint8Array.from([
      ...fixedHeader,
      ...variableHeader
    ]);
  }
}
export function parse(buffer, remainingLength, protocolVersion) {
  let pos = 0;
  const packetId = twoByteIntegerToNum(buffer, pos);
  pos += 2;
  if (protocolVersion > Mqtt.ProtocolVersion.MQTT_V3_1_1) {
    const properties = (()=>{
      const { number: length, size: consumedBytesSize } = variableByteIntegerToNum(buffer, pos);
      pos += consumedBytesSize;
      if (length > 0) {
        const prop = MqttProperties.parseMqttProperties(buffer, pos, length);
        pos += length;
        return prop;
      }
      return undefined;
    })();
    const reasonCodes = [];
    while(pos < remainingLength){
      reasonCodes.push(buffer[pos++]);
    }
    if (properties) {
      return {
        type: 'unsuback',
        packetId: packetId,
        reasonCodes: reasonCodes,
        properties: properties
      };
    } else {
      return {
        type: 'unsuback',
        packetId: packetId,
        reasonCodes: reasonCodes
      };
    }
  } else {
    return {
      type: 'unsuback',
      packetId: packetId
    };
  }
}
