// @ymjacky/mqtt5/lib/mqtt/mod.js
/**
 * This module contains constant definitions for the MQTT protocol
 * @module
 */ export const QoS = {
  AT_MOST_ONCE: 0,
  AT_LEAST_ONCE: 1,
  EXACTRY_ONCE: 2
};
export const ProtocolVersion = {
  MQTT_V3_1_1: 4,
  MQTT_V5: 5
};
export const ReasonCode = {
  Success: 0,
  NormalDisconnection: 0,
  GrantedQoS0: 0,
  GrantedQoS1: 1,
  GrantedQoS2: 2,
  DisconnectWithWillMessage: 4,
  NoMatchingSubscribers: 16,
  NoSubscriptionExisted: 17,
  ContinueAuthentication: 24,
  ReAuthenticate: 25,
  UnspecifiedError: 128,
  MalformedPacket: 129,
  ProtocolError: 130,
  ImplementationSpecificError: 131,
  UnsupportedProtocolVersion: 132,
  ClientIdentifierNotValid: 133,
  BadUserNameOrPassword: 134,
  NotAuthorized: 135,
  ServerUnavailable: 136,
  ServerBusy: 137,
  Banned: 138,
  ServerShuttingDown: 139,
  BadAuthenticationMethod: 140,
  KeepAliveTimeout: 141,
  SessionTakenOver: 142,
  TopicFilterInvalid: 143,
  TopicNameInvalid: 144,
  PacketIdentifierInUse: 145,
  PacketIdentifierNotFound: 146,
  ReceiveMaximumExceeded: 147,
  TopicAliasInvalid: 148,
  PacketTooLarge: 149,
  MessageRateTooHigh: 150,
  QuotaExceeded: 151,
  AdministrativeAction: 152,
  PayloadFormatInvalid: 153,
  RetainNotSupported: 154,
  QoSNotSupported: 155,
  UseAnotherServer: 156,
  ServerMoved: 157,
  SharedSubscriptionsNotSupported: 158,
  ConnectionRateExceeded: 159,
  MaximumConnectTime: 160,
  SubscriptionIdentifiersNotSupported: 161,
  WildcardSubscriptionsNotSupported: 162
};
export const V3_1_1_ConnectReturnCode = {
  ConnectionAccepted: 0,
  UnacceptableProtocolVersion: 1,
  IdentifierRejected: 2,
  ServerUnavailable: 3,
  BadUserNameOrPassword: 4,
  NotAuthorized: 5
};
export const V3_1_1_SubscribeReturnCode = {
  Success_MaximumQoS0: 0,
  Success_MaximumQoS1: 1,
  Success_MaximumQoS2: 2,
  Failure: 128
};
export const PacketType = {
  CONNECT: 0x1,
  CONNACK: 0x2,
  PUBLISH: 0x3,
  PUBACK: 0x4,
  PUBREC: 0x5,
  PUBREL: 0x6,
  PUBCOMP: 0x7,
  SUBSCRIBE: 0x8,
  SUBACK: 0x9,
  UNSUBSCRIBE: 0xA,
  UNSUBACK: 0xB,
  PINGREQ: 0xC,
  PINGRESP: 0xD,
  DISCONNECT: 0xE,
  AUTH: 0xF
};
export const fixexHeaderFlag = {
  CONNECT: 0b0000,
  CONNACK: 0b0000,
  PUBACK: 0b0000,
  PUBREC: 0b0000,
  PUBREL: 0b0010,
  PUBCOMP: 0b0000,
  SUBSCRIBE: 0b0010,
  SUBACK: 0b0000,
  UNSUBSCRIBE: 0b0010,
  UNSUBACK: 0b0000,
  PINGREQ: 0b0000,
  PINGRESP: 0b0000,
  DISCONNECT: 0b0000,
  AUTH: 0b0000
};
export const RetainHandling = {
  AtTheTimeOfTheSubscribe: 0,
  AtSubscribeOnlyIfTheSubscriptionDoesNotCurrentlyExist: 1,
  DoNotSend: 2
};
