// @ymjacky/mqtt5/lib/client/base_mqtt_client.js
import { Mqtt, MqttPackets } from "../mod.js";
import { Session } from "./session.js";
import { TopicAliasManager } from "./topic_alias.js";
import { ConnectTimeout, SendPacketError, StateIsNotOfflineError, StateIsNotOnlineError } from "./error.js";
import { InvalidQoSError } from "../mqtt_utils/mod.js";
import { Deferred } from "./promise.js";
import { IncomingMemoryStore } from "../mqtt_store/mod.js";
import { createCustomEvent } from "./events.js";
const defaultKeepAlive = 10; // 10 sec
const defaultConnectTimeout = 10 * 1000;
const defaultDisconnectTimeout = 1 * 1000;
const defaultPingrespTimeout = 2 * 1000;
/**
 * Abstract MQTT Client.
 * Typical functions are implemented in this class.
 */ export class BaseMqttClient {
  url;
  userName;
  password;
  clientId;
  keepAlive;
  protocolVersion;
  clean;
  connectTimeout;
  disconnectTimeout;
  connectionState;
  session;
  unresolvedConnect;
  incomingStore;
  timers;
  log;
  /**
   * ca certificate.
   * @example
   * ```
   * ["-----BEGIN CERTIFICATE-----\n...\n-----END CERTIFICATE-----\n"]
   * ```
   */ caCerts;
  /**
   * client certificate.
   * @example
   * ```
   * "-----BEGIN CERTIFICATE-----\n...\n-----END CERTIFICATE-----\n"
   * ```
   */ cert;
  /**
   * client private key.
   * @example
   * ```
   * "-----BEGIN RSA PRIVATE KEY-----\n...\n-----END RSA PRIVATE KEY-----\n"
   * ```
   */ privateKey;
  connectProperties;
  eventTarget;
  pingrespTimeoutMs;
  topicAliasManagerAboutSend;
  topicAliasMaximumAboutReceive;
  topicAliasMapAboutReceive;
  waitForClose;
  // constructor
  constructor(options){
    this.url = options?.url || new URL('mqtt://127.0.0.1:1883');
    this.log = options?.logger || (()=>{});
    this.userName = options?.username;
    this.password = options?.password;
    this.clientId = options?.clientId || '';
    this.keepAlive = (()=>{
      if (options?.keepAlive) {
        return options.keepAlive > 0 ? options.keepAlive : 0;
      } else {
        return defaultKeepAlive;
      }
    })();
    this.protocolVersion = options?.protocolVersion || Mqtt.ProtocolVersion.MQTT_V5;
    this.clean = (()=>{
      if (options) {
        return typeof options.clean === 'undefined' ? true : options.clean;
      }
      return true;
    })();
    this.connectTimeout = options?.connectTimeoutMS || defaultConnectTimeout;
    this.disconnectTimeout = options?.disconnectTimeoutMS || defaultDisconnectTimeout;
    this.connectionState = 'offline';
    this.incomingStore = options?.incomingStore || new IncomingMemoryStore();
    if (this.clean) {
      this.session = new Session(this.clientId);
    } else {
      this.session = new Session(this.clientId, options?.outgoingStore);
    }
    this.caCerts = options?.caCerts;
    this.cert = options?.cert;
    this.privateKey = options?.privateKey;
    this.eventTarget = new EventTarget();
    this.pingrespTimeoutMs = options?.pingrespTimeoutMS || defaultPingrespTimeout;
    this.timers = {};
    if (this.protocolVersion == Mqtt.ProtocolVersion.MQTT_V5 && options?.topicAliasMaximumAboutSend) {
      this.topicAliasManagerAboutSend = new TopicAliasManager(options.topicAliasMaximumAboutSend);
    }
    if (this.protocolVersion == Mqtt.ProtocolVersion.MQTT_V5 && options?.topicAliasMaximumAboutReceive) {
      this.topicAliasMaximumAboutReceive = options.topicAliasMaximumAboutReceive;
    }
    this.topicAliasMapAboutReceive = new Map();
  }
  async sendPacket(packet) {
    try {
      if (this.connectionState === 'offline') {
        this.log(`can't send packet because offline.`, packet);
        throw new SendPacketError();
      }
      this.log(`sending ${packet.type} packet`, packet);
      const bytes = MqttPackets.packetToBytes(packet, this.protocolVersion);
      this.startKeepAliveTimer();
      await this.write(bytes);
    } catch (err) {
      this.log('sendPacket: ', err);
      throw err;
    }
  }
  /**
   * Connect to Broker.
   *
   * @param options ConnectOptions
   * @returns Promise<ConnackPacket>
   * @example
   * mqtt
   * ```ts
   * const client = new MqttClient();
   * await client.connect();
   * ```
   *
   * using properties
   * ```ts
   *   const properties: MqttProperties.ConnectProperties = {
   *     sessionExpiryInterval: 300,
   *     receiveMaximum: 10,
   *     maximumPacketSize: 128,
   *     topicAliasMaximum: 20,
   *     requestResponseInformation: false,
   *     requestProblemInformation: false,
   *     userProperties: [
   *       { key: 'userProp1', val: 'userData1' },
   *       { key: 'userProp2', val: 'userData2' },
   *       { key: 'userProp2', val: 'userData3' },
   *     ],
   *   };
   *   const client = new MqttClient();
   *   await client.connect({ properties });
   * ```
   *
   * using will
   * ```ts
   *   const client = new MqttClient();
   *   const willPayload = new TextEncoder().encode('will message');
   *   await client.connect({
   *     will: {
   *       topic: 'topicA',
   *       qos: Mqtt.QoS.AT_LEAST_ONCE,
   *       payload: willPayload,
   *       properties: { userProperties: [{ key: 'userProp1', val: 'userData1' }] },
   *     },
   *   });
   * ```
   */ async connect(options) {
    if (this.connectionState != 'offline') {
      return Promise.reject(new StateIsNotOfflineError('connect'));
    }
    if (options) {
      if (typeof options.clean !== 'undefined') {
        this.clean = options.clean;
      }
      if (options.properties) {
        // save properties for reconnect
        this.connectProperties = options.properties;
      }
    }
    if (this.topicAliasMaximumAboutReceive) {
      // Tell the broker that you can receive topic alias
      if (this.connectProperties) {
        this.connectProperties.topicAliasMaximum = this.topicAliasMaximumAboutReceive;
      } else {
        this.connectProperties = {
          topicAliasMaximum: this.topicAliasMaximumAboutReceive
        };
      }
    }
    this.connectionState = 'connecting';
    const deferred = new Deferred();
    this.unresolvedConnect = deferred;
    try {
      this.startConnectTimer();
      this.log(`opening connection to ${this.url}`);
      await this.open();
      const connectPacket = {
        type: 'connect',
        clientId: this.clientId,
        username: this.userName,
        password: this.password,
        protocolVersion: this.protocolVersion,
        clean: this.clean,
        keepAlive: this.keepAlive,
        properties: this.connectProperties,
        will: options?.will
      };
      this.log(`sending ${connectPacket.type} packet`, connectPacket);
      const bytes = MqttPackets.packetToBytes(connectPacket, this.protocolVersion);
      await this.write(bytes);
    } catch (err) {
      const e = err instanceof Error ? err : err;
      this.log(`caught error opening connection: ${e.message}`);
      this.detectClosed();
      if (this.unresolvedConnect) {
        this.unresolvedConnect.reject(e);
      }
    }
    return deferred.promise;
  }
  /**
   * Returns the current number of sent publish messages that have not been completed
   */ async publishInflightCount() {
    return await this.session.publishInflightCount();
  }
  /**
   * Send a publish packet.
   * @param topic topic name
   * @param payload publish payload
   * @param options PublishOptions
   * @param properties UserPublishProperties
   * @returns Promise<PublishResult>
   * @example
   * ```ts
   * await client.publish('topicA', 'payload', { qos: Mqtt.QoS.AT_MOST_ONCE });
   * await client.publish('topicA', 'payload', { qos: Mqtt.QoS.AT_LEAST_ONCE });
   * await client.publish('topicA', 'payload', { qos: Mqtt.QoS.EXACTRY_ONCE });
   * await client.publish('topicA', new TextEncoder().encode('payloadA'), { qos: Mqtt.QoS.EXACTRY_ONCE });
   * await client.publish('topicA', 'payloadA' { qos: Mqtt.QoS.AT_LEAST_ONCE, dup: true, retain: true });
   *
   * const pubProperties: MqttProperties.UserPublishProperties = {
   *   payloadFormatIndicator: 1,
   *   messageExpiryInterval: 60,
   *   responseTopic: 'responseTopicA',
   *   correlationData: new TextEncoder().encode('correlationDataA'),
   *   userProperties: [
   *     { key: 'userProp1', val: 'userData1' },
   *     { key: 'userProp2', val: 'userData2' },
   *     { key: 'userProp2', val: 'userData3' },
   *   ],
   *   subscriptionIdentifier: 3,
   *   contentType: 'application/json',
   * };
   * await client.publish('topicA', 'payload', { qos: Mqtt.QoS.AT_MOST_ONCE }, pubProperties);
   * ```
   */ async publish(topic, payload, options, properties) {
    if (this.connectionState !== 'online') {
      throw new StateIsNotOnlineError('publish');
    }
    const dup = options && options.dup || false;
    const qos = options && options.qos || 0;
    const retain = options && options.retain || false;
    const deferred = new Deferred();
    const buffer = typeof payload === 'string' ? new TextEncoder().encode(payload) : payload;
    // resolve topic alias
    let omittedTopic;
    let topicId = 0; // 0 = not use
    if (this.topicAliasManagerAboutSend) {
      const [generated, tid] = await this.topicAliasManagerAboutSend.getTopicId(topic);
      if (!generated) {
        omittedTopic = topic;
        topic = '';
      }
      topicId = tid;
    }
    let publishProperties = properties;
    // set topic alias
    if (topicId) {
      if (publishProperties) {
        publishProperties.topicAlias = topicId;
      } else {
        publishProperties = {
          topicAlias: topicId
        };
      }
    }
    if (qos === Mqtt.QoS.AT_MOST_ONCE) {
      const packet = {
        type: 'publish',
        topic,
        payload: buffer,
        dup,
        retain,
        qos,
        packetId: 0,
        properties: publishProperties
      };
      await this.sendPacket(packet);
      deferred.resolve({
        result: 0,
        reason: ''
      });
    } else {
      const packetId = await this.session.acquirePacketId();
      const packet = {
        type: 'publish',
        topic,
        payload: buffer,
        dup,
        retain,
        qos,
        packetId,
        properties: publishProperties
      };
      await this.session.storePublish(packet, deferred, omittedTopic);
      await this.sendPacket(packet);
    }
    return deferred.promise;
  }
  /**
   * Send subscribe packets.
   */ async subscribe(input, qos, properties) {
    const arr = Array.isArray(input) ? input : [
      input
    ];
    const subs = arr.map((sub)=>{
      return typeof sub === 'object' ? {
        topicFilter: sub.topicFilter,
        qos: sub.qos || qos || Mqtt.QoS.AT_MOST_ONCE,
        retainHandling: sub.retainHandling,
        retainAsPublished: sub.retainAsPublished,
        noLocal: sub.noLocal
      } : {
        topicFilter: sub,
        qos: qos || Mqtt.QoS.AT_MOST_ONCE
      };
    });
    const deferred = new Deferred();
    const packet = {
      type: 'subscribe',
      packetId: await this.session.acquirePacketId(),
      subscriptions: subs,
      properties: properties
    };
    await this.session.storeSubscribe(packet, deferred);
    await this.sendPacket(packet);
    return deferred.promise;
  }
  /**
   * Send unsubscribe packets.
   */ async unsubscribe(input, properties) {
    const unsubs = Array.isArray(input) ? input : [
      input
    ];
    const deferred = new Deferred();
    const packet = {
      type: 'unsubscribe',
      packetId: await this.session.acquirePacketId(),
      topicFilters: unsubs,
      properties
    };
    await this.session.storeUnsubscribe(packet, deferred);
    await this.sendPacket(packet);
    return deferred.promise;
  }
  /**
   * Send disconnect packets or force disconnect.
   * Used for disconnecting.
   * Normally sends a DISCONNECT packet and waits for the Broker to disconnect.
   * In case of forced disconnection, the connection is terminated without sending a DISCONNECT packet.
   * @param force true is force disconnect, false is send disconnect packet.
   * @param properties DisconnectProperties
   * @example
   * normal disconnection
   * ```ts
   * await client.disconnect();
   * ```
   *
   * using disconnect properties.
   * ```ts
   * const properties: MqttProperties.DisconnectProperties = {
   *   reasonString: 'trouble',
   *   sessionExpiryInterval: 300,
   * };
   * await client.disconnect(false, properties);
   * ```
   * force disconnect
   * ```ts
   * await client.disconnect(true);
   * ```
   */ async disconnect(force = false, properties) {
    const deferred = new Deferred();
    this.waitForClose = deferred;
    if (this.connectionState === 'connecting') {
      this.log('A disconnect request was accepted while connected');
      if (this.unresolvedConnect) {
        await this.unresolvedConnect.promise;
      }
    }
    if (this.connectionState === 'offline') {
      deferred.resolve();
    } else {
      await this.doDisconnect(force, undefined, properties);
    }
    return deferred.promise;
  }
  async doDisconnect(force = false, reasonCode, properties) {
    if (this.connectionState === 'online') {
      if (force) {
        await this.close();
      } else {
        const disconnectPacket = {
          type: 'disconnect',
          ...reasonCode !== undefined && {
            reasonCode
          },
          ...properties && {
            properties
          }
        };
        await this.sendPacket(disconnectPacket);
        this.startDisconnectTimer();
      }
    }
  }
  /**
   * Send Auth packet.
   * @param reasonCode ReasonCode
   * @param properties AuthProperties
   */ async auth(reasonCode, properties) {
    const packet = {
      type: 'auth',
      reasonCode,
      properties
    };
    await this.sendPacket(packet);
  }
  write = (_byte)=>{};
  detectClosed() {
    this.log('connection closed.');
    this.connectionState = 'offline';
    this.stopConnectTimer();
    this.stopDisconnectTimer();
    this.stopKeepAliveTimer();
    this.stopPingrespTimer();
    if (this.waitForClose) {
      this.waitForClose.resolve();
    }
    this.eventTarget.dispatchEvent(createCustomEvent('closed', {}));
  }
  packetReceived(packet) {
    switch(packet.type){
      case 'connack':
        this.handleConnack(packet);
        break;
      case 'publish':
        this.handlePublish(packet);
        break;
      case 'puback':
        this.handlePuback(packet);
        break;
      case 'pubrec':
        this.handlePubrec(packet);
        break;
      case 'pubrel':
        this.handlePubrel(packet);
        break;
      case 'pubcomp':
        this.handlePubcomp(packet);
        break;
      case 'suback':
        this.handleSuback(packet);
        break;
      case 'unsuback':
        this.handleUnsuback(packet);
        break;
      case 'pingresp':
        this.handlePingresp(packet);
        break;
      case 'disconnect':
        this.handleDisconnect(packet);
        break;
      case 'auth':
        this.handleAuth(packet);
        break;
      default:
        throw new MqttPackets.MqttPacketsError.UnsuportedPacketError('unknown');
    }
  }
  async handleConnack(packet) {
    this.connectionState = 'online';
    this.stopConnectTimer();
    let connectionSuceeded = false;
    if (this.protocolVersion > Mqtt.ProtocolVersion.MQTT_V3_1_1) {
      if (packet.reasonCode === Mqtt.ReasonCode.Success) {
        connectionSuceeded = true;
      } else {
        this.log(`connection failed. reasonCode:${packet.reasonCode}`);
      }
    } else {
      if (packet.returnCode === Mqtt.V3_1_1_ConnectReturnCode.ConnectionAccepted) {
        connectionSuceeded = true;
      } else {
        this.log(`connection failed. reasonCode:${packet.reasonCode}`);
      }
    }
    if (connectionSuceeded) {
      // server generates clientId
      if (packet.properties?.assignedClientIdentifier) {
        this.clientId = packet.properties.assignedClientIdentifier;
      }
      if (!packet.sessionPresent) {
        this.session.clearAllStores(this.clientId);
      }
      if (packet.properties) {
        if (packet.properties.serverKeepAlive) {
          if (this.keepAlive > packet.properties.serverKeepAlive) {
            this.keepAlive = packet.properties.serverKeepAlive;
          }
        }
        if (packet.properties.topicAliasMaximum) {
          if (this.topicAliasManagerAboutSend && this.topicAliasManagerAboutSend.capacity() > packet.properties.topicAliasMaximum) {
            this.topicAliasManagerAboutSend = new TopicAliasManager(packet.properties.topicAliasMaximum);
          }
        }
      }
      // resend publish and pubrel
      for await (const packet of this.session.resendTargets()){
        this.sendPacket(packet);
      }
    }
    if (this.unresolvedConnect) {
      this.unresolvedConnect.resolve(packet);
    }
    this.eventTarget.dispatchEvent(createCustomEvent('connack', {
      detail: packet
    }));
    if (connectionSuceeded) {
      this.startKeepAliveTimer();
    }
  }
  async handlePublish(packet) {
    {
      if (packet.properties?.topicAlias) {
        if (packet.topic == '') {
          const storedTopic = this.topicAliasMapAboutReceive.get(packet.properties.topicAlias);
          if (storedTopic) {
            packet.topic = storedTopic;
          } else {
            this.log('Unregistered topic alias receiv');
            return;
          }
        } else {
          this.topicAliasMapAboutReceive.set(packet.properties.topicAlias, packet.topic);
        }
      }
    }
    if (packet.qos === Mqtt.QoS.AT_MOST_ONCE) {
      this.eventTarget.dispatchEvent(createCustomEvent('publish', {
        detail: packet
      }));
    } else if (packet.qos === Mqtt.QoS.AT_LEAST_ONCE) {
      this.sendPacket({
        type: 'puback',
        packetId: packet.packetId
      });
      this.eventTarget.dispatchEvent(createCustomEvent('publish', {
        detail: packet
      }));
    } else if (packet.qos === Mqtt.QoS.EXACTRY_ONCE) {
      const emitMessage = !packet.dup || !await this.incomingStore.has(packet.packetId);
      if (emitMessage) {
        this.incomingStore.store(packet.packetId);
        this.sendPacket({
          type: 'pubrec',
          packetId: packet.packetId
        });
        this.eventTarget.dispatchEvent(createCustomEvent('publish', {
          detail: packet
        }));
      }
    } else {
      throw new InvalidQoSError(packet.qos);
    }
  }
  async handlePuback(packet) {
    if (await this.session.packetIdInUse(packet.packetId)) {
      await this.session.discard(packet);
      const deferred = this.session.getPublishDeferred(packet.packetId);
      if (deferred) {
        if (packet.reasonCode) {
          const reason = packet.properties?.reasonString;
          deferred.resolve({
            result: packet.reasonCode,
            reason
          });
        } else {
          deferred.resolve({
            result: Mqtt.ReasonCode.Success
          });
        }
      }
    } else {
      this.log(`Unknown packetId received, it ignore. (packetId:${packet.packetId})`, 'puback');
    }
  }
  async handlePubrec(packet) {
    if (await this.session.packetIdInUse(packet.packetId)) {
      // MUST send a PUBREL packet when it receives a PUBREC packet from the receiver with a Reason Code value less than 0x80.
      // This PUBREL packet MUST contain the same Packet Identifier as the original PUBLISH packet
      // sea: https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901237
      if (packet.reasonCode && packet.reasonCode >= Mqtt.ReasonCode.UnspecifiedError // UnspecifiedError: 128
      ) {
        await this.session.discard(packet);
        const deferred = this.session.getPublishDeferred(packet.packetId);
        if (deferred) {
          const reason = packet.properties?.reasonString;
          deferred.resolve({
            result: packet.reasonCode,
            reason
          });
        }
        return;
      }
      const pubrel = {
        type: 'pubrel',
        packetId: packet.packetId
      };
      this.session.storePubrel(pubrel);
      this.sendPacket(pubrel);
    } else {
      this.log(`Unknown packetId received. (packetId:${packet.packetId})`, 'pubrec');
      if (this.protocolVersion > Mqtt.ProtocolVersion.MQTT_V3_1_1) {
        const pubrel = {
          type: 'pubrel',
          packetId: packet.packetId,
          reasonCode: Mqtt.ReasonCode.PacketIdentifierNotFound
        };
        this.sendPacket(pubrel);
      } else {
        const pubrel = {
          type: 'pubrel',
          packetId: packet.packetId
        };
        this.sendPacket(pubrel);
      }
    }
  }
  async handlePubrel(packet) {
    if (await this.incomingStore.has(packet.packetId)) {
      await this.incomingStore.discard(packet.packetId);
      this.sendPacket({
        type: 'pubcomp',
        packetId: packet.packetId
      });
    } else {
      this.log(`Unknown packetId received (packetId:${packet.packetId})`, 'pubrel');
      if (this.protocolVersion > Mqtt.ProtocolVersion.MQTT_V3_1_1) {
        const pubcomp = {
          type: 'pubcomp',
          packetId: packet.packetId,
          reasonCode: Mqtt.ReasonCode.PacketIdentifierNotFound
        };
        this.sendPacket(pubcomp);
      } else {
        const pubcomp = {
          type: 'pubcomp',
          packetId: packet.packetId
        };
        this.sendPacket(pubcomp);
      }
    }
  }
  async handlePubcomp(packet) {
    if (await this.session.packetIdInUse(packet.packetId)) {
      await this.session.discard(packet);
      const deferred = this.session.getPublishDeferred(packet.packetId);
      if (deferred) {
        if (packet.reasonCode) {
          const reason = packet.properties?.reasonString;
          deferred.resolve({
            result: packet.reasonCode,
            reason
          });
        } else {
          deferred.resolve({
            result: Mqtt.ReasonCode.Success
          });
        }
      }
    } else {
      this.log(`Unknown packetId received, it ignore. (packetId:${packet.packetId})`, 'pubcomp');
    }
  }
  async handleSuback(packet) {
    if (await this.session.packetIdInUse(packet.packetId)) {
      const rvList = packet.reasonCodes ? packet.reasonCodes : packet.returnCodes;
      await this.session.discard(packet);
      const deferred = this.session.getSubscribeDeferred(packet.packetId);
      if (deferred) {
        const reason = packet.properties?.reasonString;
        deferred.resolve({
          reasons: rvList,
          reason
        });
      }
    } else {
      this.log(`Unknown packetId received, it ignore. (packetId:${packet.packetId})`, 'suback');
    }
  }
  async handleUnsuback(packet) {
    if (await this.session.packetIdInUse(packet.packetId)) {
      await this.session.discard(packet);
      const deferred = this.session.getUnsubscribeDeferred(packet.packetId);
      if (deferred) {
        if (packet.reasonCodes) {
          const reason = packet.properties?.reasonString;
          deferred.resolve({
            reasonCodes: packet.reasonCodes,
            reason
          });
        } else {
          deferred.resolve({});
        }
      }
    } else {
      this.log(`Unknown packetId received, it ignore. (packetId:${packet.packetId})`, 'unsuback');
    }
  }
  handlePingresp(_packet) {
    this.stopPingrespTimer();
  }
  async handleDisconnect(packet) {
    await this.close();
    this.eventTarget.dispatchEvent(createCustomEvent('disconnect', {
      detail: packet
    }));
  }
  handleAuth(packet) {
    this.eventTarget.dispatchEvent(createCustomEvent('auth', {
      detail: packet
    }));
  }
  startDisconnectTimer() {
    this.startTimer('disconnect', ()=>{
      this.close();
    }, this.disconnectTimeout);
  }
  stopDisconnectTimer() {
    if (this.timerExists('disconnect')) {
      this.stopTimer('disconnect');
    }
  }
  startConnectTimer() {
    this.startTimer('connect', ()=>{
      this.close();
      if (this.unresolvedConnect) {
        this.unresolvedConnect.reject(new ConnectTimeout());
      }
    }, this.connectTimeout);
  }
  stopConnectTimer() {
    if (this.timerExists('connect')) {
      this.stopTimer('connect');
    }
  }
  startKeepAliveTimer() {
    if (!this.keepAlive) {
      return;
    }
    this.startTimer('keepAlive', async ()=>{
      await this.sendPacket({
        type: 'pingreq'
      });
      this.startPingrespTimer();
    }, this.keepAlive * 1000);
  }
  stopKeepAliveTimer() {
    if (this.timerExists('keepAlive')) {
      this.stopTimer('keepAlive');
    }
  }
  startPingrespTimer() {
    this.startTimer('pingrespTimeout', async ()=>{
      await this.disconnect(true);
    }, this.pingrespTimeoutMs);
  }
  stopPingrespTimer() {
    if (this.timerExists('pingrespTimeout')) {
      this.stopTimer('pingrespTimeout');
    }
  }
  startTimer(name, cb, delay) {
    if (this.timerExists(name)) {
      this.stopTimer(name);
    }
    this.timers[name] = setTimeout(()=>{
      delete this.timers[name];
      cb();
    }, delay);
  }
  stopTimer(name) {
    if (!this.timerExists(name)) {
      return;
    }
    const timerId = this.timers[name];
    if (timerId) {
      clearTimeout(timerId);
      delete this.timers[name];
    }
  }
  timerExists(name) {
    return !!this.timers[name];
  }
  /**
   * @param type
   * @param callback
   * @example
   * Wait for receipt of a connack packet.
   * ```ts
   * client.on('connack', (event) => {
   *   const connackPacket = event.detail;
   *   console.log(connackPacket);
   * });
   * ```
   *
   * Wait for receipt of a publish message.
   * ```ts
   * client.on('publish', (event) => {
   *   const packet = event.detail;
   *   const receiveMessage = decoder.decode(packet.payload);
   *   console.log(`topic: ${packet.topic}`, `message: ${receiveMessage}`);
   * });
   * ```
   *
   * Wait for events when disconnected from the broker or disconnected by itself.
   * ```ts
   * client.on('closed', () => {
   *   console.log('closed');
   * });
   * ```
   *
   * Waiting to receive Disconnect packets sent by the Broker
   * ```ts
   * client.on('disconnect', (event) => {
   *   const disconnectPacket = event.detail;
   *   console.log(disconnectPacket.reasonCode);
   * });
   * ```
   * wait for Auth packet.
   * ```ts
   * client.on('auth', (event) => {
   *   const authPacket = event.detail;
   *   console.log(authPacket);
   * });
   * ```
   */ on = (type, callback)=>{
    this.eventTarget.addEventListener(type, callback);
  };
  off = (type, callback)=>{
    this.eventTarget.removeEventListener(type, callback);
  };
  /**
   * return clientId.
   * @returns clientId
   */ getClientId() {
    return this.clientId;
  }
}
