// @ymjacky/mqtt5/lib/client/error.js
export class ConnectionClosed extends Error {
  #className = 'ConnectionClosed';
  constructor(){
    super(`connection closed`);
  }
}
export class ConnectionReset extends Error {
  #className = 'ConnectionReset';
  constructor(message){
    super(message);
  }
}
export class ConnectTimeout extends Error {
  #className = 'ConnectTimeout';
  constructor(){
    super('connect timed out');
  }
}
export class StateIsNotOfflineError extends Error {
  #className = 'StateIsNotOfflineError';
  constructor(message){
    super(`State is not offline: ${message}`);
  }
}
export class StateIsNotOnlineError extends Error {
  #className = 'StateIsNotOnlineError';
  constructor(message){
    super(`State is not online: ${message}`);
  }
}
export class SendPacketError extends Error {
  #className = 'SendPacketError';
  constructor(){
    super(`can't send packet because offline`);
  }
}
export class UnexpectedUrlProtocol extends Error {
  #className = 'UnexpectedUrlProtocolextends';
  constructor(message){
    super(`unexpected URL protcol: ${message}`);
  }
}
