// @ymjacky/mqtt5/lib/client/session.js
import { IdProvider } from "../mod.js";
import { MemoryStore } from "../mqtt_store/mod.js";
export class Session {
  sessionId;
  packetIdProvider;
  inflightPublish;
  inflightSubscribe;
  inflightUnsubscribe;
  outgoingStore;
  constructor(sessionId, outgoingStore){
    this.sessionId = sessionId;
    this.packetIdProvider = new IdProvider.ReuseIdProvider(1, 65535); // packetId range is 1 to 65535
    this.outgoingStore = outgoingStore || new MemoryStore();
    this.inflightPublish = new Map();
    this.inflightSubscribe = new Map();
    this.inflightUnsubscribe = new Map();
  }
  getSessionId() {
    return this.sessionId;
  }
  async acquirePacketId() {
    return await this.packetIdProvider.acquire();
  }
  async clearAllStores(newSessionId) {
    if (newSessionId) {
      this.sessionId = newSessionId;
    }
    await this.outgoingStore.clear();
    this.inflightPublish.clear();
    this.inflightSubscribe.clear();
    this.inflightUnsubscribe.clear();
    this.packetIdProvider.clear();
  }
  async packetIdInUse(packetId) {
    return await this.packetIdProvider.inUse(packetId);
  }
  async discard(packet) {
    await this.packetIdProvider.release(packet.packetId);
    if (packet.type === 'puback' || packet.type === 'pubrec' || packet.type === 'pubcomp') {
      await this.outgoingStore.discard(packet.packetId);
    }
  }
  async *resendTargets() {
    for await (const packet of this.outgoingStore.resendIterator()){
      if (packet.type === 'publish') {
        packet.dup = true;
      }
      if (packet.packetId) {
        this.packetIdProvider.registerIfNotInUse(packet.packetId);
      }
      yield packet;
    }
  }
  async storePublish(packet, deferred, omittedTopic) {
    this.inflightPublish.set(packet.packetId, deferred);
    if (omittedTopic) {
      const copiedPacket = {
        ...packet
      };
      if (packet.properties) {
        copiedPacket.properties = {
          ...packet.properties
        };
        delete copiedPacket?.properties?.topicAlias;
      }
      // Restore omitted topics and then register them
      copiedPacket.topic = omittedTopic;
      return await this.outgoingStore.store(copiedPacket);
    } else {
      return await this.outgoingStore.store(packet);
    }
  }
  async storePubrel(packet) {
    return await this.outgoingStore.store(packet);
  }
  getPublishDeferred(packetId) {
    const deferred = this.inflightPublish.get(packetId);
    this.inflightPublish.delete(packetId);
    return deferred;
  }
  storeSubscribe(packet, deferred) {
    this.inflightSubscribe.set(packet.packetId, deferred);
  }
  getSubscribeDeferred(packetId) {
    const deferred = this.inflightSubscribe.get(packetId);
    this.inflightSubscribe.delete(packetId);
    return deferred;
  }
  storeUnsubscribe(packet, deferred) {
    this.inflightUnsubscribe.set(packet.packetId, deferred);
  }
  getUnsubscribeDeferred(packetId) {
    const deferred = this.inflightUnsubscribe.get(packetId);
    this.inflightUnsubscribe.delete(packetId);
    return deferred;
  }
  async publishInflightCount() {
    return await this.outgoingStore.count();
  }
}
