// @ymjacky/mqtt5/lib/client/topic_alias.js
import { LruCache } from "../cache/mod.js";
import { ReuseIdProvider } from "../id_provider/mod.js";
import { Deferred } from "./promise.js";
export class TopicAliasManager {
  topicIdProvider;
  topicAliasMap;
  deferred;
  constructor(topicAliasMaximum){
    this.topicIdProvider = new ReuseIdProvider(1, topicAliasMaximum);
    this.topicAliasMap = new LruCache(topicAliasMaximum);
  }
  capacity() {
    return this.topicAliasMap.capacity();
  }
  isFull() {
    return this.topicAliasMap.size() == this.topicAliasMap.capacity();
  }
  async getTopicId(topic) {
    if (this.deferred) {
      await this.deferred.promise;
    } else {
      this.deferred = new Deferred();
    }
    let topicId = this.topicAliasMap.get(topic);
    let generated = false;
    if (!topicId) {
      topicId = await this.registerTopic(topic);
      generated = true;
    }
    this.deferred.resolve();
    return [
      generated,
      topicId
    ];
  }
  async registerTopic(topic) {
    if (this.topicAliasMap.has(topic)) {
      const tid = this.topicAliasMap.get(topic);
      if (tid) {
        this.topicIdProvider.release(tid);
      }
      const topicId = await this.acquireTopicId();
      this.topicAliasMap.set(topic, topicId);
      return topicId;
    } else {
      if (this.isFull()) {
        const lruTopic = this.topicAliasMap.lruKey();
        if (lruTopic) {
          const tid = this.topicAliasMap.get(lruTopic);
          if (tid) {
            this.topicIdProvider.release(tid);
          }
          this.topicAliasMap.delete(lruTopic);
        }
      }
      const topicId = await this.acquireTopicId();
      this.topicAliasMap.set(topic, topicId);
      return topicId;
    }
  }
  releaseTopic(topic) {
    if (this.topicAliasMap.has(topic)) {
      const topicId = this.topicAliasMap.get(topic);
      if (topicId) {
        this.topicIdProvider.release(topicId);
        this.topicAliasMap.delete(topic);
      }
    }
  }
  async acquireTopicId() {
    return await this.topicIdProvider.acquire();
  }
}
