// @ymjacky/mqtt5/lib/id_provider/reuse_id_provider.js
export class ReuseIdProvider {
  _start;
  _end;
  _use = [];
  _reuse = [];
  _reusable = [];
  _next;
  constructor(start, end){
    if (start >= end) {
      throw new Error('Invalid argument');
    }
    this._start = start;
    this._end = end;
    this._use = [];
    this._reuse = [];
    this._reusable = [];
    this._next = this._start;
  }
  clear() {
    this._use = [];
    this._reuse = [];
    this._reusable = [];
    this._next = this._start;
  }
  acquire() {
    let id = this._reusable.shift(); // remove first entry
    if (id) {
      this._reuse.push(id);
    } else {
      if (this._next > this._end) {
        throw new Error('Unable to obtain ID because the limit has been reached');
      }
      id = this._next++;
      this._use.push(id);
    }
    return id;
  }
  release(id) {
    let index = this._reuse.findIndex((v)=>v === id);
    if (index >= 0) {
      this._reuse.splice(index, 1); // delete
      if (id === this._next - 1) {
        this._next = id;
      } else {
        this._reusable.push(id);
      }
      return true;
    }
    index = this._use.findIndex((v)=>v === id);
    if (index >= 0) {
      this._use.splice(index, 1); // delete
      if (id === this._next - 1) {
        this._next = id;
      } else {
        this._reusable.push(id);
      }
      return true;
    }
    return false;
  }
  inUse(id) {
    return this._reuse.includes(id) || this._use.includes(id);
  }
  registerIfNotInUse(id) {
    if (!this.inUse(id)) {
      if (this._next <= id) {
        this._next = id + 1;
      }
      this._use.push(id);
      return true;
    }
    return false;
  }
}
