// @ymjacky/mqtt5/lib/mqtt_store/store.js
export class IncomingMemoryStore {
  packets = new Set();
  store(packetId) {
    this.packets.add(packetId);
    return Promise.resolve();
  }
  has(packetId) {
    return Promise.resolve(this.packets.has(packetId));
  }
  discard(packetId) {
    this.packets.delete(packetId);
    return Promise.resolve();
  }
}
export class MemoryStore {
  packets = new Map();
  store(packet) {
    if (!packet.packetId) {
      return Promise.reject(new Error('missing packet.packetId'));
    }
    this.packets.set(packet.packetId, packet);
    return Promise.resolve();
  }
  has(packetId) {
    const exist = this.packets.has(packetId);
    return Promise.resolve(exist);
  }
  discard(packetId) {
    this.packets.delete(packetId);
    return Promise.resolve();
  }
  count() {
    return Promise.resolve(this.packets.size);
  }
  clear() {
    this.packets.clear();
    return Promise.resolve();
  }
  async *resendIterator() {
    for (const value of this.packets.values()){
      yield value;
    }
  }
}
